bootstrap/updater_test.go

```diff
@@ -0,0 +1,418 @@
+	require.EqualValues(t, 1, got.ID)
+	require.Len(t, got.Data, 2)
+	require.EqualValues(t, got.Data[0].Epoch, 0)
+	require.EqualValues(t, "0x6fe7c971", got.Data[0].Beacon.String())
+	require.Empty(t, got.Data[0].ActiveSet)
+	require.EqualValues(t, got.Data[1].Epoch, 1)
+	require.EqualValues(t, "0x6fe7c971", got.Data[1].Beacon.String())
+	require.Len(t, got.Data[1].ActiveSet, 2)
+	require.Equal(t, types.HexToHash32("85de8823d6a0cd251aa62ce9315459302ea31ce9701531d3677ac8ba548a4210"), got.Data[1].ActiveSet[0].Hash32())
+	require.Equal(t, types.HexToHash32("65af4350d28f3d953c6c6660e37954698839125fbda7aac3edcef469b2ad9e64"), got.Data[1].ActiveSet[1].Hash32())
+}
+
+func checkUpdate2(t *testing.T, got *bootstrap.VerifiedUpdate) {
+	require.EqualValues(t, 2, got.ID)
+	require.Len(t, got.Data, 2)
+	require.EqualValues(t, got.Data[0].Epoch, 1)
+	require.EqualValues(t, "0x6fe7c971", got.Data[0].Beacon.String())
+	require.Len(t, got.Data[0].ActiveSet, 2)
+	require.Equal(t, types.HexToHash32("85de8823d6a0cd251aa62ce9315459302ea31ce9701531d3677ac8ba548a4210"), got.Data[0].ActiveSet[0].Hash32())
+	require.Equal(t, types.HexToHash32("65af4350d28f3d953c6c6660e37954698839125fbda7aac3edcef469b2ad9e64"), got.Data[0].ActiveSet[1].Hash32())
+	require.EqualValues(t, got.Data[1].Epoch, 2)
+	require.EqualValues(t, "0xf70cf90b", got.Data[1].Beacon.String())
+	require.Len(t, got.Data[1].ActiveSet, 2)
+	require.Equal(t, types.HexToHash32("0575fc4083eb5b5c4422063c87071eb5123d4db6fee7bc1ecb02e52e97916aef"), got.Data[1].ActiveSet[0].Hash32())
+	require.Equal(t, types.HexToHash32("23716e2667034edc62595a6d1628ff5c323cf099f2cc161e5653a96c9fd2bd55"), got.Data[1].ActiveSet[1].Hash32())
+}
+
+func checkUpdate3(t *testing.T, got *bootstrap.VerifiedUpdate) {
+	require.EqualValues(t, 3, got.ID)
+	require.Len(t, got.Data, 2)
+	require.EqualValues(t, got.Data[0].Epoch, 2)
+	require.EqualValues(t, "0xf70cf90b", got.Data[0].Beacon.String())
+	require.Len(t, got.Data[0].ActiveSet, 2)
+	require.Equal(t, types.HexToHash32("0575fc4083eb5b5c4422063c87071eb5123d4db6fee7bc1ecb02e52e97916aef"), got.Data[0].ActiveSet[0].Hash32())
+	require.Equal(t, types.HexToHash32("23716e2667034edc62595a6d1628ff5c323cf099f2cc161e5653a96c9fd2bd55"), got.Data[0].ActiveSet[1].Hash32())
+	require.EqualValues(t, got.Data[1].Epoch, 3)
+	require.EqualValues(t, "0x9ef76b65", got.Data[1].Beacon.String())
+	require.Len(t, got.Data[1].ActiveSet, 3)
+	require.Equal(t, types.HexToHash32("65af4350d28f3d953c6c6660e37954698839125fbda7aac3edcef469b2ad9e64"), got.Data[1].ActiveSet[0].Hash32())
+	require.Equal(t, types.HexToHash32("e46b23d64140357b16d18eace600b28ab767bfd7b51c8e9977a342b71c3a23dd"), got.Data[1].ActiveSet[1].Hash32())
+	require.Equal(t, types.HexToHash32("85de8823d6a0cd251aa62ce9315459302ea31ce9701531d3677ac8ba548a4210"), got.Data[1].ActiveSet[2].Hash32())
+}
+
+type checkFunc func(*testing.T, *bootstrap.VerifiedUpdate)
+
+func TestNew(t *testing.T) {
+	tcs := []struct {
```

Comment:
can you elaborate please how does it make more it difficult? you can run any test case from this list

Reply:
I should have clarified, I'm referring to debugging with vscode, a code lens provides a single click mechanism to debug a test, which runs the whole test.

If you want to debug a specific subtest you can either place your breakpoint and skip past all the subtests you are not interested in, manually make a debug config for the specific subtest and use that or comment out the entries for the other subtests. They are all more work than placing your breakpoint on the test you are interested in and clicking on the code lens.

Reply:
do you mean that in vscode you can not invoke a subtests independently?

in goland if i can just place a breakpoint in the main shared code and invoke the subtest to debug.

my main objection to the approach you are suggesting is that there can be code in btwn multiple invocation of `t.Run()` that makes the tests not hermetic. and my experience, different from yours, is that this approach almost always result in more code.

Reply:
@countvonzero yes, invoking individual table driven subtests  is [not currently supported for vscode](https://github.com/golang/vscode-go/issues/1602)

With regard to the hermeticness, I guess that is up to the test writer, maybe its useful to run some code between invocations, in this case it doesn't seem necessary to run code between invocations.

With regard to code size I thought I'd just see what happens with this test you can see the result [here](https://github.com/spacemeshos/go-spacemesh/blob/a10abcb13d4790a1d028e553bae6c8e798b6a3a3/bootstrap/updater_test.go#L146-L195) in this case it seems to have reduced test size by about 15% I also find the test slightly easier to read like this, but maybe that is just me.

Reply:
thanks for offering the code.

is this a deal breaker for you? if not i'd rather keep the test as hermetic as possible.

Reply:
Not a dealbreaker, but I prefer the version I linked due to (smaller size, easier debug for vscode and I think easier to read), so I don't see a reason not to use it, except for the hermeticness which I don't feel I've really understood. Could you give an example or point out in the [linked code](https://github.com/spacemeshos/go-spacemesh/blob/a10abcb13d4790a1d028e553bae6c8e798b6a3a3/bootstrap/updater_test.go#L146-L195) what are the problems that it poses with regards to hermeticness?

Reply:
in your code (quoted below), you could insert code for the next t.Run(), but then had to clean it for the second t.Run()
now only the `persisted` map matter, but maintenance or future changes, it may not be the case.

basically before every invocation one needs to worry about whether the state is set correctly. in the current code, each subtest is completely separate. i feel going forward it's less error-prone.

for me ease of maintenance is more important than curtailing the code to specific IDE needs.

```
func TestNew(t *testing.T) {
	t.Run("no recovery", func(t *testing.T) { testNew(t, nil, nil) })

	persisted := map[string]string{"00001-2023-03-18T22-26-13": update1}
	t.Run("recovery one", func(t *testing.T) { testNew(t, persisted, checkUpdate1) })

	persisted = map[string]string{
		"00014-2023-04-18T22-26-11": update3,
		"00001-2023-03-18T22-26-13": update1,
		"00002-2023-04-18T22-26-13": update2,
	}
	t.Run("recovery latest", func(t *testing.T) { testNew(t, persisted, checkUpdate3) })
}
```

Reply:
Thanks for the explanation @countvonzero I think I understand. I have come across the problem you mention in the past, generally when re-using complex test setups to test multiple cases. In this scenario given the simplicity of the the test parameters I don't think it's a significant concern. In terms of hermeticness I think I would rate the two variants equally, in the variant I provided state could be messed up between calls to t.Run, but in the current version it seems to me that the same could happen by reusing values across multiple iterations of the for loop. Anyway I'm happy to leave it as it is.

Reply:
my guiding principal is: with little cost, make code future proof whenever i can. so the two approaches are actually different for me.

if you had done
```go
 func TestNew(t *testing.T) {
	t.Run("no recovery", func(t *testing.T) { testNoRecovery(t)})
	t.Run("recovery", func(t *testing.T) { testRecovery(t)})
}
```
where there are no dependencies btwn sub tests, then i'll definitely oblige. like i said, this approach usually resulted in more code.